bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java

```diff
@@ -683,9 +744,73 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
 
 	@Override
 	public void setVisibleRegion(int start, int length) {
-		fWasProjectionEnabled= isProjectionMode();
-		disableProjection();
-		super.setVisibleRegion(start, length);
+		if (!isProjectionMode()) {
+			super.setVisibleRegion(start, length);
+			return;
+		}
+		IDocument document= getDocument();
+		if (document == null) {
+			return;
+		}
+		try {
+			// If the visible region changes, make sure collapsed regions outside of the old visible regions are expanded
+			// and collapse everything outside the new visible region
+			int end= computeEndOfVisibleRegion(start, length, document);
+			expandOutsideCurrentVisibleRegion(document);
+			collapseOutsideOfNewVisibleRegion(start, end, document);
+			fVisibleRegionDuringProjection= new Region(start, end - start);
+		} catch (BadLocationException e) {
+			String PLUGIN_ID= "org.eclipse.jface.text"; //$NON-NLS-1$
+			Bundle bundle= Platform.getBundle(PLUGIN_ID);
+			ILog log= ILog.of(bundle);
+			log.log(new Status(IStatus.WARNING, getClass(), IStatus.OK, e.getMessage(), e));
+		}
+		fVisibleRegionDuringProjection= new Region(start, length);
```

Comment:
is this line meant to be inside the catch block? Or why it overrides the value set in the try block?

Reply:
That line shouldn't be there at all - removed.